allennlp/models/segmental_language_model.py

```diff
@@ -0,0 +1,300 @@
+        """
+        if hasattr(self._contextualizer, 'num_layers') and hasattr(self._backward_segmental_contextualizer, 'num_layers'):
+            return self._contextualizer.num_layers + self._backward_segmental_contextualizer.num_layers + 1
+        else:
+            raise NotImplementedError(f"Contextualizer of type {type(self._contextualizer)} " +
+                                      "does not report how many layers it has.")
+
+    def forward(self,  # type: ignore
+                tokens: Dict[str, torch.LongTensor],
+                tags: torch.Tensor,
+                seg_map: torch.Tensor,
+                seg_ends: torch.Tensor,
+                seg_starts: torch.Tensor) -> Dict[str, torch.Tensor]:
+        """
+        Computes the averaged forward (and backward, if language model is bidirectional)
+        LM loss from the batch.
+
+        By convention, the input dict is required to have at least a ``"tokens"``
+        entry that's the output of a ``SingleIdTokenIndexer``, which is used
+        to compute the language model targets.
+
+        Parameters
+        ----------
+        tokens: ``torch.Tensor``, required.
+            The output of ``Batch.as_tensor_dict()`` for a batch of sentences.
+
+        Returns
+        -------
+        Dict with keys:
+
+        ``'loss'``: ``torch.Tensor``
+            forward negative log likelihood, or the average of forward/backward
+            if language model is bidirectional
+        ``'forward_loss'``: ``torch.Tensor``
+            forward direction negative log likelihood
+        ``'backward_loss'``: ``torch.Tensor`` or ``None``
+            backward direction negative log likelihood. If language model is not
+            bidirectional, this is ``None``.
+        ``'lm_embeddings'``: ``Union[torch.Tensor, List[torch.Tensor]]``
+            (batch_size, timesteps, embed_dim) tensor of top layer contextual representations or
+            list of all layers. No dropout applied.
+        ``'noncontextual_token_embeddings'``: ``torch.Tensor``
+            (batch_size, timesteps, token_embed_dim) tensor of bottom layer noncontextual
+            representations
+        ``'mask'``: ``torch.Tensor``
+            (batch_size, timesteps) mask for the embeddings
+        """
+        # pylint: disable=arguments-differ
+        mask = get_text_field_mask(tokens)
+
+        # shape (batch_size, timesteps, embedding_size)
+        embeddings = self._text_field_embedder(tokens)
+
+        # Either the top layer or all layers.
+        contextual_embeddings: Union[torch.Tensor, List[torch.Tensor]] = self._contextualizer(
+                embeddings, mask
+        )
+
+        return_dict = {'lm_embeddings': contextual_embeddings,
+                       'noncontextual_token_embeddings': embeddings,
+                       'mask': mask
+                       }
+
+        token_ids = tokens.get("tokens")
+        if token_ids is None:
+            return return_dict
```

Comment:
Wait, so there isn't any segmental info in the case where the LM is being used as an embedder? I think I need to plead for one more overview of the entire system tomorrow!